Following camera Focal Point offset

tomaszrybakiewicz opened the pull request:
Closes https://github.com/mapbox/mapbox-navigation-android/issues/4347

## Changelog entry
```
- Added `FollowingFrameOptions.focalPoint` that can be used to define the position of the first framed geometry point (typically the user location indicator, if available) in the `MapboxNavigationViewportDataSource.followingPadding`. [#5875](https://github.com/mapbox/mapbox-navigation-android/pull/5875)
```

## TODOs
- [x] bump Maps SDK dependency
- [x] add changelog entry

tomaszrybakiewicz commented:
@LukasPaczos What do you think about this approach for offsetting the following camera focal point?

abhishek1508 commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSourceOptions.kt:
```diff
@@ -44,6 +44,13 @@ class FollowingFrameOptions internal constructor() {
     var maxZoom = 16.35
 
+    /**
+     * Focal point for following camera frames.
```
```suggestion
     * Defines focal point for camera frames when in FOLLOWING mode with a non-zero pitch. The focal point remains centered when the camera pitch is 0 in FOLLOWING mode.
```

LukasPaczos replied:
Pitch has no direct impact on the calculation. Only the relationship with `maximizeViewableGeometryWhenPitchZero` is relevant, more on that in the other suggestion.

abhishek1508 reviewed (commented):
This LGTM. It's a simple implementation and adding the focal point to the options allows the users to change it anytime they would want to.

LukasPaczos commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSourceOptions.kt:
```diff
@@ -44,6 +44,13 @@ class FollowingFrameOptions internal constructor() {
      */
     var maxZoom = 16.35
 
+    /**
+     * Focal point for following camera frames.
+     *
+     * Defaults to `FocalPoint(0.5, 1.0)` that centers horizontally on frame bottom edge.
+     */
+    var focalPoint: FocalPoint = FocalPoint(0.5, 1.0)
```
```suggestion
    /**
     * Focal point that defines the position of the first framed geometry point (typically the user location indicator, if available)
     * in the [MapboxNavigationViewportDataSource.followingPadding].
     *
     * The value is a horizontal and vertical ratio starting from the top left corner of the padding, in the `<0.0, 1.0>` range.
     * For example:
     * - FocalPoint(0.0, 0.0)` is positions the first geometry point in the top left
     * - FocalPoint(0.5, 0.5)` is positions the first geometry point in the center
     * - FocalPoint(1.0, 1.0)` is positions the first geometry point in the bottom right
     *
     * Defaults to `FocalPoint(0.5, 1.0)` that centers horizontally on the bottom edge of the padding.
     *
     * **NOTE:** The focal point change has no effect when the camera's pitch is `0` and [maximizeViewableGeometryWhenPitchZero] is enabled.
     */
    var focalPoint: FocalPoint = FocalPoint(0.5, 1.0)
```

LukasPaczos replied:
Additionally, let's add a note to `maximizeViewableGeometryWhenPitchZero`:
```
**NOTE:** When [maximizeViewableGeometryWhenPitchZero] is enabled and the camera pitch is `0`, the [focalPoint] value is ignored.
```

LukasPaczos commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/FocalPoint.kt:
```diff
@@ -0,0 +1,6 @@
+package com.mapbox.navigation.ui.maps.camera.data
+
+data class FocalPoint(
```
How about nesting this type in `FollowingFrameOptions`, `FollowingFrameOptions.FocalPoint`? `FocalPoint` is rather generic and has a pretty narrow application on our camera APIs.

tomaszrybakiewicz replied:
Sounds good to me.

LukasPaczos commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/FocalPoint.kt:
```diff
@@ -0,0 +1,6 @@
+package com.mapbox.navigation.ui.maps.camera.data
+
+data class FocalPoint(
+    val x: Double,
+    val y: Double
```
Let's check that the values are in 0 to 1 range and throw an exception if not.

tomaszrybakiewicz replied:
Values less than 0.0 and greater than 1.0 are also acceptable to allow puck placement outside the padded frame.

```
FocalPoint(0.5, 1.1) - will place the puck below the bottom edge of the frame (centered horizontally)
FocalPoint(0.5, -0.9) - will place the puck above the top edge of the frame (centered horizontally)
FocalPoint(-0.9, 0.5) - will place the puck to the left of the left edge of the frame (centered vertically)
FocalPoint(1.1, 0.5) - will place the puck to the right of the right edge of the frame (centered vertically)
```

<img width="418" alt="Screen Shot 2022-05-30 at 10 41 06 AM" src="https://user-images.githubusercontent.com/2678039/171015335-6d21a1bc-a23a-4bd9-a03e-89490ad407eb.png">

LukasPaczos replied:
When the focal point (and effectively the center of the camera) is below the padding, the whole `ViewportDataSource` logic breaks down. The class' task is to fit the provided geometry (route plus puck) into the frame (padding). When we put the focal point outside of the frame, we have no possibility to fit the geometry into the frame anymore.

In the video above, we run into the specific safeguard of the GL-Native API where geometries below the bottom edge of the padding are not framed (that's why error logs are not printed). If you would try the same if any other edge, we'd be falling back to previous valid zoom level values, but this doesn't change the fact that it makes no sense from the perspective of the `ViewportDataSource`. For example:
![Screenshot_20220530_182544_small](https://user-images.githubusercontent.com/16925074/171031728-9632f84f-eec4-49ad-b07c-a073c76f7ddf.png)
In this screenshot (`FocalPoitn(1.2, 0.8)`) we just remain in some previous fallback zoom level while not being able to frame any of the geometry (and GL-Native prints error logs on each attempt).

LukasPaczos reviewed (commented):
Looks like no technical blockers on the maps side, great!

LukasPaczos commented:
Having some fun with the new feature :upside_down_face: 

https://user-images.githubusercontent.com/16925074/171010324-498974e7-4881-4593-8359-d14d013d802c.mp4

LukasPaczos commented:
Unfortunately, the safeguard that was added in GL-Native because the puck was designed to sit at the bottom edge doesn't help us now (refs https://github.com/mapbox/mapbox-navigation-android/pull/5875#discussion_r884976501). Current attempt to unlock to position of the puck leads to the safeguard being more of a problem.

For example, experience of driving to the north with camera bearing locked to `0`:

https://user-images.githubusercontent.com/16925074/171030994-52c7bcc8-6ad8-46be-b3ab-430835892eaf.mp4

everything works well, we frame until we reach maximum zoom level.

However, when we try driving to the south with camera bearing locked to `0`:

https://user-images.githubusercontent.com/16925074/171031158-5602f0ee-1c43-451e-b623-2c0a1085d759.mp4

things break down because whatever is below the focal point of the camera becomes ignored. In this video that's basically all of the geometry and we just fallback to max zoom level value.

LukasPaczos commented:
I'm not sure when changes will land upstream (maybe only in 10.8) but the changes for Nav SDK look good to me.

tomaszrybakiewicz commented:
I've rebased on top of the latest `main`, added unit-tests and regenerated Metalava API file.
I moved SNAPSHOT testing to a separate branch [tr-#4347-following-camera-focal-point_testing](https://github.com/mapbox/mapbox-navigation-android/compare/tr-%234347-following-camera-focal-point...tr-%234347-following-camera-focal-point_testing) to avoid accidental merge of `build.gradle` file with SNAPSHOT changes.

I am also adding a list of TODO items to finish before we can merge this upstream.

tomaszrybakiewicz commented:
@LukasPaczos This is finally unblocked and ready to go. I've rebased on top of latest `main` with Maps v10.7.0 and added CHANGELOG entry. I leave it up to you to decide in which NavSDK version you want to ship this.

LukasPaczos reviewed (approved):
🚢 

The feature is well contained so let's ship it in `v2.7.0-rc.1` this week.
